tools/sstable_consumer.hh

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "seastarx.hh"
+
+#include <seastar/core/loop.hh>
+
+namespace sstables {
+class sstable;
+}
+
+class partition_start;
+class static_row;
+class clustering_row;
+class range_tombstone_change;
+class partition_end;
+
+// stop_iteration::no -> continue consuming sstable content
+class sstable_consumer {
+public:
+    virtual ~sstable_consumer() = default;
+    // called at the very start
+    virtual future<> on_start_of_stream() = 0;
+    // stop_iteration::yes -> on_end_of_sstable() - skip sstable content
+    // sstable parameter is nullptr when merging multiple sstables
+    virtual future<stop_iteration> on_new_sstable(const sstables::sstable* const) = 0;
+    // stop_iteration::yes -> consume(partition_end) - skip partition content
+    virtual future<stop_iteration> consume(partition_start&&) = 0;
+    // stop_iteration::yes -> consume(partition_end) - skip remaining partition content
+    virtual future<stop_iteration> consume(static_row&&) = 0;
+    // stop_iteration::yes -> consume(partition_end) - skip remaining partition content
+    virtual future<stop_iteration> consume(clustering_row&&) = 0;
+    // stop_iteration::yes -> consume(partition_end) - skip remaining partition content
+    virtual future<stop_iteration> consume(range_tombstone_change&&) = 0;
+    // stop_iteration::yes -> on_end_of_sstable() - skip remaining partitions in sstable
+    virtual future<stop_iteration> consume(partition_end&&) = 0;
+    // stop_iteration::yes -> full stop - skip remaining sstables
+    virtual future<stop_iteration> on_end_of_sstable() = 0;
+    // called at the very end
+    virtual future<> on_end_of_stream() = 0;
+};
```

Comment:
I wonder if we should use the reader interface instead.

Reply:
We use the reader consumer interface + some additional methods on top as we are consuming a stream consisting of many sstables. A consumer (push) interface is simpler to write as all the logic of feeding the consumer is in a single place, in the infrastructure level. This was the original motivation for implementing all sstable operations as consumers and this decision was carried over to the script operation as well.
A reader interface (pull) is equally viable but leads to more boilerplate as all consumer have to have have their own consume loop.

Reply:
We have a reader consumer too. Anyway, I don't see a big benefit to switch. Maybe we can script pipeline operations one day (read many sstables -> combining_reader -> script).

Reply:
This is already supported via the `--merge` flag. When used, all sstable readers will be joined in a combined reader and then this combined stream being fed to the stream. The `scylla sstable dump-data` also supports this flag.